src/network/networkd-radv.c

```diff
@@ -21,9 +21,138 @@
 #include <arpa/inet.h>
 
 #include "networkd-address.h"
+#include "networkd-manager.h"
 #include "networkd-radv.h"
 #include "sd-radv.h"
 
+static int radv_get_ip6dns(Network *network, struct in6_addr **dns,
+                           size_t *n_dns) {
+        _cleanup_free_ struct in6_addr *addresses = NULL;
+        size_t i, n_addresses = 0, n_allocated = 0;
+
+        assert(network);
+        assert(dns);
+        assert(n_dns);
+
+        for (i = 0; i < network->n_dns; i++) {
+                union in_addr_union addr;
+
+                if (network->dns[i].family != AF_INET6)
+                        continue;
+
+                addr.in6 = network->dns[i].address.in6;
+
+                if (in_addr_is_null(AF_INET6, &addr) ||
+                    in_addr_is_link_local(AF_INET6, &addr) ||
+                    in_addr_is_localhost(AF_INET6, &addr))
+                        continue;
+
+                if (!GREEDY_REALLOC(addresses, n_allocated, n_addresses + 1))
+                        return log_oom();
+
+                addresses[n_addresses++] = addr.in6;
+        }
+
+        if (addresses) {
+                free(*dns);
+                *dns = addresses;
+                addresses = NULL;
+
+                *n_dns = n_addresses;
+        }
+
+        return n_addresses;
+}
+
+static int radv_set_dns(Link *link, Link *uplink) {
+        _cleanup_free_ struct in6_addr *dns = NULL;
+        size_t n_dns;
+        usec_t lifetime_usec;
+        int r;
+
+        if (!link->network->router_emit_dns)
+                return 0;
+
+        if (link->network->router_dns) {
+                dns = memdup(link->network->router_dns,
+                             link->network->n_router_dns * sizeof(struct in6_addr));
+                n_dns = link->network->n_router_dns;
+                lifetime_usec = link->network->router_dns_lifetime_usec;
+
+                goto set_dns;
+        }
+
+        lifetime_usec = SD_RADV_DEFAULT_DNS_LIFETIME_USEC;
+
+        r = radv_get_ip6dns(link->network, &dns, &n_dns);
+        if (r > 0)
+                goto set_dns;
+
+        if (uplink) {
+                r = radv_get_ip6dns(uplink->network, &dns, &n_dns);
```

Comment:
hmm, so here the uplink DNS info overrides the locally configured one. Does that make sense? shouldn't it be the other way round? i.e. if there's local config use it, otherwise find uplink?

Reply:
No, uplink should not override here, if the current link has sensible DNS servers, they get set and this is skipped by the previous goto set_dns.